dm/pkg/checker/table_structure.go

```diff
@@ -60,71 +74,128 @@ func (o *incompatibilityOption) String() string {
+	c := &TablesChecker{
 		db:     db,
 		dbinfo: dbinfo,
 		tables: tables,
 	}
+	return c
 }
 
 // Check implements RealChecker interface.
-func (c *TablesChecker) Check(ctx context.Context) *Result {
+func (c *TablesChecker) Check(ctx context.Context) (*Result, error) {
 	r := &Result{
 		Name:  c.Name(),
 		Desc:  "check compatibility of table structure",
 		State: StateSuccess,
 		Extra: fmt.Sprintf("address of db instance - %s:%d", c.dbinfo.Host, c.dbinfo.Port),
 	}
 
+	startTime := time.Now()
+
+	concurrency := len(c.tables)/tablesForOneThread + 1
+	if concurrency > maxThreadNum {
+		concurrency = maxThreadNum
+	}
+	maxConnections, err := utils.GetMaxConnections(ctx, c.db)
+	if err != nil {
+		return r, err
+	}
+	if concurrency > maxConnections {
+		concurrency = maxConnections
+	}
+
 	var (
-		err        error
-		options    = make(map[string][]*incompatibilityOption)
-		statements = make(map[string]string)
+		optCh   = make(chan *incompatibilityOption, MaxOptions*len(c.tables))
+		errCh   = make(chan error, 1)
+		inCh    = make(chan *filter.Table, concurrency)
+		checkWg sync.WaitGroup
 	)
-	for schema, tables := range c.tables {
-		if len(tables) == 0 {
-			tables, err = dbutil.GetTables(ctx, c.db, schema)
-			if err != nil {
-				markCheckError(r, err)
-				return r
+
+	checkCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	go func() {
+		defer close(inCh)
+		for _, table := range c.tables {
+			select {
+			case inCh <- table:
+			case <-checkCtx.Done():
+				return
 			}
 		}
-
-		for _, table := range tables {
-			tableName := dbutil.TableName(schema, table)
-			statement, err := dbutil.GetCreateTableSQL(ctx, c.db, schema, table)
-			if err != nil {
-				// continue if table was deleted when checking
-				if isMySQLError(err, mysql.ErrNoSuchTable) {
-					continue
+	}()
+	log.L().Logger.Info("start to check tables", zap.Int("concurrency", concurrency), zap.Int("all table num", len(c.tables)))
+	checkFunc := func() {
+		defer checkWg.Done()
+		for {
+			select {
+			case <-checkCtx.Done():
+				return
+			case table, ok := <-inCh:
+				if !ok {
+					return
+				}
+				statement, err := dbutil.GetCreateTableSQL(ctx, c.db, table.Schema, table.Name)
+				if err != nil {
+					// continue if table was deleted when checking
+					if isMySQLError(err, mysql.ErrNoSuchTable) {
+						continue
+					}
+					if len(errCh) == 0 {
+						errCh <- err
+					}
+					break
 				}
-				markCheckError(r, err)
-				return r
-			}
 
-			opts := c.checkCreateSQL(ctx, statement)
-			if len(opts) > 0 {
-				options[tableName] = opts
-				statements[tableName] = statement
+				opts := c.checkCreateSQL(ctx, statement)
+				for _, opt := range opts {
+					opt.tableID = table.String()
+					optCh <- opt
+				}
 			}
 		}
 	}
 
-	for name, opts := range options {
-		if len(opts) == 0 {
-			continue
+	go func() {
+		defer cancel()
+		err, ok := <-errCh
+		if ok {
+			markCheckError(r, err)
+			close(errCh)
```

Comment:
we should only close a channel in a sender goroutine if the sender is the only sender of the channel see more [here](https://go101.org/article/channel-closing.html)